pgpubsub/channel.py

```diff
@@ -147,50 +148,61 @@ class TriggerChannel(BaseChannel):
     @classmethod
     def deserialize(cls, payload: Union[Dict, str]):
         payload_dict = super().deserialize(payload)
-        old_model_data = cls._build_model_serializer_data(payload_dict, state='old')
-        new_model_data = cls._build_model_serializer_data(payload_dict, state='new')
-
-        old_deserialized_objects = serializers.deserialize(
-            'json',
-            json.dumps(old_model_data, cls=DjangoJSONEncoder),
-            ignorenonexistent=True,
-        )
-        new_deserialized_objects = serializers.deserialize(
-            'json',
-            json.dumps(new_model_data, cls=DjangoJSONEncoder),
-            ignorenonexistent=True,
-        )
-
-        old = next(old_deserialized_objects, None)
-        if old is not None:
-            old = old.object
-        new = next(new_deserialized_objects, None)
-        if new is not None:
-            new = new.object
+        old = cls._deserialize_from_state(payload_dict, state='old')
+        new = cls._deserialize_from_state(payload_dict, state='new')
         return {'old': old, 'new': new}
 
     @classmethod
-    def _build_model_serializer_data(cls, payload: Dict, state: str):
-        """Reformat serialized data into shape as expected
-        by the Django model deserializer.
-        """
+    def _is_up_to_date(cls, app: str, db_version_id: int) -> bool:
+        """Check if the db version id from django migrations is the latest for the given app"""
+        newer_migration_exists = MigrationRecorder.Migration.objects.filter(
+            app=app, id__gt=db_version_id
+        ).exists()
+        return not newer_migration_exists
+
+    @classmethod
+    def _deserialize_from_state(cls, payload: Dict, state: str) -> Any:
         app = payload['app']
         model_name = payload['model']
         model_cls = apps.get_model(
             app_label=payload['app'],
             model_name=payload['model'],
         )
-        fields = {
-            field.name: field for field in model_cls._meta.fields
-        }
-        db_fields = {
-            field.db_column: field for field in model_cls._meta.fields
-        }
-
-        original_state = payload[state]
+        db_version = payload.get("db_version", None)
+        if db_version is None or cls._is_up_to_date(app, db_version):
+            model_data = cls._build_model_serializer_data(model_cls, payload[state])
+
+            deserialized_objects = serializers.deserialize(
+                'json',
+                json.dumps(model_data, cls=DjangoJSONEncoder),
+                ignorenonexistent=True,
+            )
+
+            obj = next(deserialized_objects, None)
+            if obj is not None:
+                obj = obj.object
+        else:
+            if payload[state]:
+                try:
+                    obj = model_cls.objects.get(pk=payload[state][model_cls._meta.pk.name])
+                except model_cls.DoesNotExist:
+                    obj = None
```

Comment:
The fact that we don't get the values that were stored during the modification is just an inherent downside of this solution. The initial problem and other possible solutions to it are described in https://github.com/Opus10/django-pgpubsub/issues/38

This solution is quite easy to implement and rather straighforward as compared to other solutions. IMHO this solution is a good default behavior. The existing behavour would be in the best case that Notification is skipped and in the worst case the silent data corruption in the downstream consumers of the notifications. If you think this is not a good default behaviour I can put it under the configuration option and make it off by default.

Regarding the `old` vs `new` in the creation scenario I'll add a test and make sure that old is always `None` for this use case.

Reply:
Also I can add a flag like `outdated_replaced_with_latest` to the listener which would indicate that the serialized form was changed and the latest version was read from DB. This would however break the compatibility of the existing listener functions as they do not expect this parameter. I guess I can pass it if the listener signature has it.

Reply:
> I'd expect the old value of a listener to be None, but with this change (if the migration history had changed), old and new would both be the same.

I've added the test that shows that this is the behaviour that the current code implements.

Reply:
> Part of the reason to store the JSON was to be sure to have a snapshot of what the triggering instance looked like at the time of the triggering event.

I understand that design decision but what about the issues described in https://github.com/Opus10/django-pgpubsub/issues/38. Using JSON and make the consumer responsible for schema changes would be an option if the listener received JSON, but the listener receives the django model and it is the library responsibility to make sure it is consistent. What would be the proposed way to address the issue?

Reply:
I think this is a really clever solution - using the migration recorder is something I'd never have thought of. I am worried, though, that the departure in behaviour (as in my above comment) between this and the regular flow could lead to even more confusing bugs than the bugs which occur due to the current implementation as discussed in #38. The idea of explicitly declaring somewhere via the `outdated_replaced_with_latest` flag would definitely go some way to address this, since that would make it clear to the dev to expect different behaviour from the library in the event a migration occurs. I think this would perhaps make more sense to declare at the channel level, rather than the listener, since it is the channel that's facilitating the transformation to the new entity?

Of all the solutions listed on #38, I actually think the idea of using django-pghistory could be the neatest, but probably the biggest change. I've actually never used pghistory (Wes' library), but I have used simple-history quite extensively. I would assume pghistory respects migrations though in the same way that simple-history does. The advtange of pghistory is of course that it can store snapshots at the postgres level through triggers, so we could potentially generate the snapshot in the same trigger that generates the JSON - that would actually be a pretty beautiful collaboration of these libraries.

One challenge with this solution though would be how to do it in a compatible way with the current JSON field - would this solution completely replace that?

I'll have more of a think over the next few days about the best approach to this. What do you think though of your history suggestion? Do you think it seems too much work?

Reply:
>  this would perhaps make more sense to declare at the channel level, rather than the listener

I think you misunderstood what I'm proposing here. `outdated_replaced_with_latest` is a flag passed to the listener that signals that the outdated serialized entity was received and that the latest entity was passed to the listener. Earlier in the thread I've proposed to have another option that should specify if this detection of the outdate entity and reading of the latest one should be performed at all (my message about the default behaviour). This can be global with a possibility to override in the channel.

>  how to do it in a compatible way with the current JSON field - would this solution completely replace that?
I was thinking about the complete replacement. I can't see how they can coexists so that it makes sense.

> we probably would need to think about how to hook that into the migration system when the main model is generated
I'm not familiar how django-pghistory works. I expect that it already has this functionality to generate migrations for history table based on changes to the main model. If it is not used and cannot be adapted to store snapshots on demand I guess that functionality needs to be replicated in this pgpubsub.

>  What do you think though of your history suggestion? Do you think it seems too much work?
As I'm not familiar with django-pghistory I can't really say. Also I can't invest much time into the project.

Reply:
What about this for an idea?

We package the library with a base class `Channeled` (best name TBD). `Channeled` represents a general type which can be passed in to a listener as an argument. It will basically be a thin wrapper around the model types which already appear as listener arguments, adorned with some special attributes we wish to add.

So, the initial version would look like this:

```
@dataclass
class Channeled:
    outdated_replaced_with_latest: bool
    instance: models.Model
```

If we take the `Media` example, we'd then ask the client to define a subclass

```
@dataclass
class ChanneledMedia(Channeled):
    instance: Media
```
The media listener would then consume the `ChanneledMedia` type, rather than the `Media` type:

```
@pgpubsub.post_save_listener(MediaTriggerChannel)
def scan_media(old: ChanneledMedia, new: ChanneledMedia):
    ...

```

It's then up to the channel to set the `outdated_replaced_with_latest` and `instance` on the `old` and `new` variables before passing through to the listener. We would of course still need to support regular models as listener argument types - if these are used, we just don't consider issue #38.

What do you think?

Reply:
I'm not sure I understand what you try to solve by this. It can't see scenario when we can make use of `old` version which has `outdated_replaced_with_latest==True` as it will be the same as `new`. 

Maybe I didn't mention explicitly that when the db was migrated `old` should be `None`. So for it makes more sense to have `outdated_replaced_with_latest` just as a parameter to the listener. If you don't like that we add parameters that describes what is happening we may have more generic parameter that contains this kind of metadata (e.g it may also contain an indicator if it is insertion or update) with name like `notification_metadata`.

I can see your point that the approach with reading the latest version may not be acceptable in every scenario. Initially I though that it might solve most cases but now I need to think and look to the use cases I have at hand.

Reply:
Yeah, the more I think about it, the more I think just grabbing the new version is too much of a departure from the normal behaviour, even if the `outdated_replaced_with_latest` is passed. I think it would also be quite difficult for a developer to know what to do if `outdated_replaced_with_latest` is True - I think it makes things almost more confusing/worse than the notification just failing because it's outdated.

What about if we give the option to supply default deserialisation values/overrides on the channel, so that these default values are used when deserialising the notification data to the model instance for the listener. For example:

```
@dataclass
class MediaTriggerChannel(TriggerChannel):
    model = Media
    lock_notifications = True
    deserialization_defaults: dict = field(default_factory=lambda: {'size': 0})
```

I'm not sure if the `deserialization_defaults` API is optimal, but the basic idea is to allow specifying custom deserialisation.